### ISSUE TITLE
MP-Config: Failing integration tests for custom `ConfigSource` implementations

### ISSUE THREAD

**Issue by fwippe** [quarkusio/quarkus#20190]
`@QuarkusTest` integration tests run by the `maven-failsafe-plugin` are failing when being run in projects with custom `ConfigSource` implementations. The error message is:

```
java.util.ServiceConfigurationError: org.eclipse.microprofile.config.spi.ConfigSource: x.y.CustomConfigSource not a subtype
```

Note that the same tests run successfully when being initiated by the `maven-surefire-plugin`.

Reproducer:

```java
public class CustomConfigSource implements ConfigSource {
    static final String TEST_CUSTOM_PROPERTY = "test.custom.property";

    @Override
    public Set<String> getPropertyNames() {
        return Set.of(TEST_CUSTOM_PROPERTY);
    }

    @Override
    public String getValue(String propertyName) {
        return TEST_CUSTOM_PROPERTY.equals(propertyName) ? "success" : null;
    }

    @Override
    public String getName() {
        return "custom-test";
    }
}

@QuarkusTest
public class CustomConfigSourceIT {
    @ConfigProperty(name = CustomConfigSource.TEST_CUSTOM_PROPERTY)
    String testProperty;

    @Test
    public void mustInjectCustomProperty() {
        assertThat(testProperty, is("success"));
    }
}
```

See https://github.com/fwippe/custom-config-source for a reproducing project (just run `./mvnw verify`).

**Comment by geoand** [comment]
cc @radcortez

**Comment by fwippe** [comment]
Update: Work-around: Downgrading failsafe-plugin to version 2.22.2 resolved the issue.

**Comment by ismailsimsek** [comment]
for me it fails with quarkus 2.2.X and 2.1.X and works with 2.0.X

**Comment by Naros** [comment]
Unfortunately downgrading `failsafe` is not an option for Debezium.

**Comment by radcortez** [comment]
Usually, you get that exception when there is a mismatch on the classloaders. I'm not sure what failsafe is doing differently from one version to another. @stuartwdouglas any idea?

**Comment by stuartwdouglas** [comment]
So setting: `<useModulePath>false</useModulePath>` will fix this.

It seems failsafe is generating a manifest with a module name, and it is breaking everything. We used to have a workaround for module issues where we would return null in the ClassLoader getParent(), we may need to bring this back as modular environments are very problematic in terms of controlling service loader delegation.

**Comment by Naros** [comment]
@stuartwdouglas @radcortez So after some investigating to our issues migrating to Quarkus 2.1+ from 2.0 it seems for us the issue is with `ConfigProvider#getConfig()`.  When we get the `Config` object from this method, the priority of resolving a given configuration property has seemly changed in precedence somehow.

In our project we have one artifact called `debezium-server` that has its own test suite that defines a `ConfigSource` that is loaded through the ServiceLoader.  We then have another artifact that depends on the server artifact called `debezium-server-kafka` and it too defines its own `ConfigSource` but this source implementation extends the server's source class and overrides the values in the configuration map.

Now in 2.0.0.Final, the following lines resulted in the property from `debezium-server-kafka` configuration map being resolved:
```
final Config config = ConfigProvider.getConfig();
final String name = config.getValue("debezium.sink.type", String.class);
```
However after upgrading to 2.2.3.Final, the same code resolves the property from the `debezium-server` configuration map.  When I look at the sources that make up the configuration, I do see source from `debezium-server-kafka` listed too but unfortunately it seems the priority precendence has somehow changes.

Is this somehow related to this or is this separate by chance?

**Comment by radcortez** [comment]
@Naros can you please provide me the ordinal and names of the sources? These may influence the order of the sources. We did some changes that may affect the ordering is both the ordinals and names are the same.

**Comment by Naros** [comment]
Hi @radcortez, so I checked and regardless of the version both have the same ordinal of `100` and name of `test`; however what I do notice is that their order in `ConfigProvider.getConfig().getSources()` is reversed.

In 2.0.0.Final the `debezium-server-kafka` configuration comes first and resolves `debezium.sink.type` as `kafka`.
In 2.2.3.Final the `debezium-server` configuration comes first and resolves `debezium.sink.type` as `test`.

Is there a workaround for this?

**Comment by radcortez** [comment]
> I'm guessing the workaround for now would be to explicitly implement `getOrdinal()` in the kafka artifact and return an ordinal higher than the default to address the configuration priority change or to lower the ordinal in the server artifact's configuration. Was this compatibility change intended or a regression?

Correct.

Well, a change that we did was to remove the runtime ServiceLoader discovery and turn it into a build time discovery. I'm guessing that the intermediate Sets change the old order and because both of your sources have the same sorting values (ordinal and then name), you get them in an unwanted order.

Even if we didn't change anything, this could happen. We know that in the majority of the cases the CP and ServiceLoader keep their order, but there is a small % of executions where the order change. So, I really recommend setting a specific ordinal to make sure you get the right order.

**Comment by Naros** [comment]
That makes sense @radcortez and thanks for the explanation.  I'll adjust our code accordingly.

**Comment by Naros** [comment]
In an effort to upgrade to Quarkus 2.4.0.CR1 we now get the specific ConfigSource is not a subtype error.

**Comment by radcortez** [comment]
> In an effort to upgrade to Quarkus 2.4.0.CR1 we now get the specific ConfigSource is not a subtype error.

Is this because of of the Failsafe plugin version?

**Comment by fwippe** [comment]
Closed along with #22770
